src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/MetadataUtilsSmi.cs

```diff
@@ -0,0 +1,625 @@
+                        extendedCode = DetermineExtendedTypeCode(value);
+
+                        // Some types aren't allowed for Variants but are for the general-purpos function.  
+                        //  Match behavior of other types and return invalid in these cases.
+                        if (ExtendedClrTypeCode.SqlXml == extendedCode)
+                        {
+                            extendedCode = ExtendedClrTypeCode.Invalid;
+                        }
+                        break;
+                    case SqlDbType.Udt:
+                        throw ADP.DbTypeNotSupported(SqlDbType.Udt.ToString());
+                    case SqlDbType.Time:
+                        if (value.GetType() == typeof(TimeSpan))
+                            extendedCode = ExtendedClrTypeCode.TimeSpan;
+                        break;
+                    case SqlDbType.DateTimeOffset:
+                        if (value.GetType() == typeof(DateTimeOffset))
+                            extendedCode = ExtendedClrTypeCode.DateTimeOffset;
+                        break;
+                    case SqlDbType.Xml:
+                        if (value.GetType() == typeof(SqlXml))
+                            extendedCode = ExtendedClrTypeCode.SqlXml;
+                        if (value.GetType() == typeof(XmlDataFeed))
+                            extendedCode = ExtendedClrTypeCode.XmlReader;
+                        else if (value.GetType() == typeof(System.String))
+                            extendedCode = ExtendedClrTypeCode.String;
+                        break;
+                    case SqlDbType.Structured:
+                        if (isMultiValued)
+                        {
+                            if (value is IEnumerable<SqlDataRecord>)
+                            {
+                                extendedCode = ExtendedClrTypeCode.IEnumerableOfSqlDataRecord;
+                            }
+                            else if (value is DbDataReader)
+                            {
+                                extendedCode = ExtendedClrTypeCode.DbDataReader;
+                            }
+                        }
+                        break;
+                    default:
+                        // Leave as invalid
+                        break;
+                }
+            }
+
+            return extendedCode;
+        }
+
+        // Method to map from Type to ExtendedTypeCode
+        static internal ExtendedClrTypeCode DetermineExtendedTypeCodeFromType(Type clrType)
+        {
+            object result = s_typeToExtendedTypeCodeMap[clrType];
+
+            ExtendedClrTypeCode resultCode;
+            if (null == result)
+            {
+                resultCode = ExtendedClrTypeCode.Invalid;
+            }
+            else
+            {
+                resultCode = (ExtendedClrTypeCode)result;
+            }
+
+            return resultCode;
```

Comment:
For that matter is the type of `s_typeToExtendedTypeCodeMap` appears to be a .NET legacy. It could be a `Dictionary<Type, ExtendedClrTypeCode>` and then:

    ExtendedClrTypeCode result;
    return s_typeToExtendedTypeCodeMap.TryGetValue(clrType, out result) ? result : ExtendedClrTypeCode.Invalid;

Would avoid a cast.